File: ProblemSolving/450DSA/JavaScript/src/arrays/MinimumSwapsAndKTogether.js
/**
 * @author Anirudh Sharma
 * 
 * Given an array of n positive integers and a number k.
 * 
 * Find the minimum number of swaps required to bring all the numbers less than or equal to k together.
 * 
 * Constraints:
 * 1 ≤ N ≤ 10^5
 * 1 ≤ nums[i], K ≤10^7
 */
const minimumSwaps = (nums, k) => {
    // Special cases
    if (nums === undefined || nums.length < 2) {
        return 0;
    }
    // Count elements which are less than or equal to k
    let lessThanOrEqualToK = 0;
    for (let i = 0; i < nums.length; i++) {
        if (nums[i] <= k) {
            lessThanOrEqualToK++;
        }
    }
    // Count number of elements which are greater than k
    // int the first window of size lessThanOrEqualToK
    let greaterThanOrEqualToK = 0;
    for (let i = 0; i < lessThanOrEqualToK; i++) {
        if (nums[i] > k) {
            greaterThanOrEqualToK++;
        }
    }
    // Final count of swaps required
    let swaps = greaterThanOrEqualToK;
    // Using two pointer technique
    let left = 0;
    let right = lessThanOrEqualToK;
    // Loop until right pointer reaches to the end
    while (right < nums.length) {
        // Decrement count of previous window
        if (nums[left] > k) {
            --greaterThanOrEqualToK;
        }
        // Increment count of current window
        if (nums[right] > k) {
            ++greaterThanOrEqualToK;
        }
        left++;
        right++;
        swaps = Math.min(swaps, greaterThanOrEqualToK);
    }
    return swaps;
};


let arr = [2, 1, 5, 6, 3];
let K = 3;
console.log(minimumSwaps(arr, K));

arr = [2, 7, 9, 5, 8, 7, 4];
K = 6;
console.log(minimumSwaps(arr, K));

File: ProblemSolving/450DSA/Python/src/arrays/MinimumSwapsAndKTogether.py
"""
@author Anirudh Sharma

Given an array of n positive integers and a number k.

Find the minimum number of swaps required to bring all the numbers less than or equal to k together.

Constraints:
1 ≤ N ≤ 10^5
1 ≤ nums[i], K ≤10^7
"""


def minimumSwaps(nums, k):
    # Special cases
    if nums is None or len(nums) < 2:
        return 0
    # Count of elements less than or equal to k
    lessThanOrEqualToK = 0
    for n in nums:
        if n <= k:
            lessThanOrEqualToK += 1
    # Count of elements greather than k in window
    # of size lessThanOrEqualToK
    greaterThanK = 0
    for i in range(0, lessThanOrEqualToK):
        if nums[i] > k:
            greaterThanK += 1
    # Minimum number if swaps required
    swaps = greaterThanK
    # Left and right pointers for sliding window
    left = 0
    right = lessThanOrEqualToK
    # Loop until right pointer reaches to the end of the array
    while right < len(nums):
        # Decrement count of previous window
        if nums[left] > k:
            greaterThanK -= 1
        # Increment count of current window
        if nums[right] > k:
            greaterThanK += 1
        # Minimize number of swaps required
        left += 1
        right += 1
        swaps = min(swaps, greaterThanK)
    return swaps


if __name__ == "__main__":
    nums = [2, 1, 5, 6, 3]
    k = 3
    print(minimumSwaps(nums, k))

    nums = [2, 7, 9, 5, 8, 7, 4]
    k = 6
    print(minimumSwaps(nums, k))
